### PR TITLE
Catalog payments from data_base::push

### PR DESCRIPTION
**Pull request by kulpreet**
Includes a small refactor of block exists tests to allow independent tests for cataloged payments without checking other block related conditions.

### CODE REVIEW

File: src/data_base.cpp [hunk 1]
```diff
@@ -519,6 +518,9 @@ code data_base::push(const block& block, size_t height,
     if (!blocks_->validate(block.hash(), error::success))
         return error::operation_failed;
 
+    if (catalog(block))
```
[the remark this answers is not in the document]
Comment (kulpreet) [reply src/data_base.cpp]: Yeah. I had the same problem of suddenly testing for a positive outcome to observe failure.

I have changed this to `if (catalog(block) != error::success)` to streamline it with the other calls that are noting failures by a false return.

File: src/databases/address_database.cpp [hunk 2]
```diff
@@ -142,7 +142,8 @@ void address_database::store(const hash_digest& script_hash, const point& point,
 // Confirmation of payment is dynamically derived from current tx state.
 void address_database::catalog(const transaction& tx)
 {
-    BITCOIN_ASSERT(tx.metadata.state);
+    BITCOIN_ASSERT(tx.metadata.link);
+    BITCOIN_ASSERT(!tx.metadata.existed);
```
[the remark this answers is not in the document]
Comment (kulpreet) [reply src/databases/address_database.cpp]: Yikes! Thanks.

File: src/data_base.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (kulpreet) [reply src/data_base.cpp]: Switched to `if((ec = operation))` pattern being used everywhere else instead.

File: test/data_base.cpp [hunk 3]
```diff
@@ -32,10 +32,72 @@ using namespace bc::database;
 using namespace boost::system;
 using namespace boost::filesystem;
 
+static void test_outputs_cataloged(const address_database& payments_store,
+    const transaction& tx, bool is_found)
+{
+    for (size_t j = 0; j < tx.outputs().size(); ++j)
+    {
+        const auto& output = tx.outputs()[j];
+        output_point outpoint{ tx.hash(), static_cast<uint32_t>(j) };
```
[the remark this answers is not in the document]
Comment (kulpreet) [reply test/data_base.cpp]: Changed type, and removed the cast.

**Comment by kulpreet** [comment]
@evoskuil - updated the PR with fixes and changes.

File: test/data_base.cpp [hunk 4]
```diff
@@ -63,150 +116,45 @@ static void test_block_exists(const data_base& interface, size_t height,
...
     BOOST_REQUIRE(!is_confirmed(result.state()));
 
-    for (size_t i = 0; i < block0.transactions().size(); ++i)
+    uint32_t tx_position = 0u;
+
+    for (const auto& tx: block.transactions())
     {
-        const auto& tx = block0.transactions()[i];
-        const auto tx_hash = tx.hash();
+        if (!catalog)
```
Comment (evoskuil) [inline test/data_base.cpp]: Why is this in the loop, it’s value cannot change in the loop?

File: test/data_base.cpp [hunk 5]
```diff
@@ -63,150 +116,45 @@ static void test_block_exists(const data_base& interface, size_t height,
...
-                //         {
-                //             found = true;
-                //             break;
-                //         }
-                //     }
-
-                //     BOOST_REQUIRE(!found);
-                // }
-            }
+            test_inputs_cataloged(payments_store, tx, true);
```
Comment (evoskuil) [inline test/data_base.cpp]: Shouldn’t catalog (vs true) be passed here (and below)?

File: src/data_base.cpp [hunk 6]
```diff
@@ -519,6 +520,10 @@ code data_base::push(const block& block, size_t height,
     if (!blocks_->validate(block.hash(), error::success))
         return error::operation_failed;
 
+    if ((ec = catalog(block))) {
+        return error::operation_failed;
```
[the remark this answers is not in the document]
Comment (kulpreet) [reply src/data_base.cpp]: OK, makes sense. Changed this to propagate the error.

I fell into the trap of keeping the error aligned with the others by returning an operation_failed error.

File: test/data_base.cpp (hunk 4 above)
Comment (kulpreet) [reply test/data_base.cpp]: Moved.

File: test/data_base.cpp (hunk 5 above)
Comment (kulpreet) [reply test/data_base.cpp]: I took out the `catalog` parameter, as I want these helpers to be called only if cataloging is enabled, so the `catalog` parameter is redundant.

The true|false instead captures if the payment should be found in the address database (aka payment store here). Recall I renamed a badly named argument from `is_found` to `expect_found`. This true|false is for that argument.

File: test/data_base.cpp [hunk 7]
```diff
@@ -32,10 +32,62 @@ using namespace bc::database;
...
 
+static void test_outputs_cataloged(const address_database& payments_store,
+    const transaction& tx, bool expect_found)
+{
+    uint32_t output_index = 0u;
+
+    for (const auto& output: tx.outputs())
+    {
+        output_point outpoint{ tx.hash(), output_index };
+        const auto& script_hash = sha256_hash(output.script().to_data(false));
```
[the remark this answers is not in the document]
Comment (kulpreet) [reply test/data_base.cpp]: Changed, also did a pass on the whole file to spot any other occurrences of this.

Just as a note, I was running with the idea of that assigning rvalues to const refs is safe. I agree though, it doesn't help and can only cause potential confusion in the future.

[the remark this answers is not in the document]
Comment (evoskuil) [reply test/data_base.cpp]: It’s actually not safe. The rvalue can and will at unpredictable times be taken out of scope by the compiler despite the reference.

[the remark this answers is not in the document]
Comment (kulpreet) [reply test/data_base.cpp]: Thanks for catching these. I'll avoid such assignments in future.

I have also made a pass through this file and changed other such assignments.

**Comment by kulpreet** [comment]
@evoskuil - Updated PR with changes.

File: src/data_base.cpp [hunk 8]
```diff
@@ -521,7 +521,7 @@ code data_base::push(const block& block, size_t height,
         return error::operation_failed;
 
     if ((ec = catalog(block))) {
-        return error::operation_failed;
+        return ec;
```
Comment (evoskuil) [inline src/data_base.cpp]: Braces are unbalanced. Braces should not be used for single line scope.

**Pull request merged**